### PR TITLE
Changes dependency from fftw@3 to fftw-api@3.

### PR DESCRIPTION
**Pull request by dipietrantonio**
Also, updates the conflict clause for table locking.

### CODE REVIEW

File: var/spack/repos/builtin/packages/casacore/package.py [hunk 1]
```diff
@@ -53,8 +53,8 @@ class Casacore(CMakePackage):
     depends_on("lapack")
     depends_on("cfitsio")
     depends_on("wcslib@4.20:+cfitsio")
-    depends_on("fftw@3.0.0: precision=float,double", when="@3.4.0:")
-    depends_on("fftw@3.0.0: precision=float,double", when="~fftpack")
+    depends_on("fftw-api@3 precision=float,double", when="@3.4.0:")
+    depends_on("fftw-api@3 precision=float,double", when="~fftpack")
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/casacore/package.py:57]: This is wrong, in the sense that virtual dependencies do not have variant attached at the moment. Can you add back the additional constraint on `fftw`, if needed, and clean the one on the virtual dependency?

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/casacore/package.py (hunk 1 above)
Comment (pelahi) [reply var/spack/repos/builtin/packages/casacore/package.py:57]: Hi @alalazo, is this documented somewhere? I will say that this worked. And to be clear, you are saying have something like this?
```python
    depends_on("fftw-api@3", when="@3.4.0:")
    depends_on("fftw-api@3", when="~fftpack")
    depends_on("fftw precision=float,double", when="~fftpack")
```

Comment (alalazo) [reply var/spack/repos/builtin/packages/casacore/package.py:57]: @pelahi To be fair, it is badly documented at https://spack.readthedocs.io/en/latest/packaging_guide.html#virtual-dependencies but we usually stress this part in tutorials. 

>  I will say that this worked.

I'm pretty sure this was just interpreted as:
```python
depends_on("fftw-api@3", when="@3.4.0:")
depends_on("fftw-api@3", when="~fftpack")
```

> And to be clear, you are saying have something like this?

Almost. This:
```python
depends_on("fftw-api@3", when="@3.4.0:")
depends_on("fftw-api@3", when="~fftpack")
depends_on("fftw precision=float,double", when="^fftw")
```
should work and allow you to pick any other `fftw-api` provider even if `~fftpack` (assuming that makes sense).

File: var/spack/repos/builtin/packages/casacore/package.py [hunk 2]
```diff
@@ -66,7 +66,7 @@ class Casacore(CMakePackage):
     depends_on("gsl", when="+dysco")
 
     conflicts("~mpi", when="+adios2")
-    conflicts("+tablelocking", when="+mpi")
+    conflicts("+tablelocking", when="+mpi +adios2")
```
Comment (mpokorny) [inline var/spack/repos/builtin/packages/casacore/package.py:69]: Enabling MPI, https://github.com/casacore/casacore/blob/5a8df94738bdc36be27e695d7b14fe949a1cc2df/CMakeLists.txt#L420-L429, is a sufficient condition to prohibit the table locking option, regardless of the value of the adios2 option. Additionally, casacore allows building against adios2 without MPI enabled (in adios2) https://github.com/casacore/casacore/blob/5a8df94738bdc36be27e695d7b14fe949a1cc2df/CMakeLists.txt#L328-L337 . Not having tried all the combinations, I cannot verify they all actually work in casacore, and perhaps I've missed something, but this change doesn't agree with what's in the CMakeLists.txt file. Is there some particular configuration that you're trying to prohibit that isn't captured by the current conflicts?

Comment (dipietrantonio) [reply var/spack/repos/builtin/packages/casacore/package.py:69]: @pelahi, can you comment on this as you are the author of the change?

Comment (dipietrantonio) [reply var/spack/repos/builtin/packages/casacore/package.py:69]: @pelahi ?

Comment (alalazo) [reply var/spack/repos/builtin/packages/casacore/package.py:69]: @dipietrantonio Are you still interested in the PR? Can you take care of review comments?

Comment (dipietrantonio) [reply var/spack/repos/builtin/packages/casacore/package.py:69]: @alalazo I am withdrawing the PR as it seems this is something we better keep locally. Thank you for your insightful comments nonetheless.

**Pull request closed without merging**